fix: UTC-299: Dont use project members for agreement selected annotators

hakan458 opened the pull request:
<!-- Follow Merge -->
This PR was automatically generated via Follow Merge. [Workflow Run](https://github.com/HumanSignal/label-studio/actions/runs/18014691703)
Please ensure that all linked upstream Pull Requests are merged before proceeding with this one.
- https://github.com/HumanSignal/label-studio-sdk/pull/622
<!-- Follow Merge -->







Adding endpoint to get unique list of user IDs that have actually submitted an annotation in a project. 
Will be used to populated agreement_selected.annotators (currently using all project members)

Commit:
fix: UTC-299: Dont use org members for agreement selected annotators

Commit:
update all urls

Commit:
return simpleuser from annotators api

Commit:
add migration for existing views

Commit:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-utc-299

Commit:
fix migration deps

Commit:
fix urls

jombooth commented on label_studio/data_manager/migrations/0016_migrate_agreement_selected_annotators_to_unique.py:
```diff
@@ -0,0 +1,98 @@
+    if not created:
+        return  # already in progress or done
+
+    # Look up models at runtime inside the worker process
+    View = django_apps.get_model('data_manager', 'View')
+    Annotation = django_apps.get_model('tasks', 'Annotation')
```
🤯 awesome

jombooth commented on label_studio/data_manager/migrations/0016_migrate_agreement_selected_annotators_to_unique.py:
```diff
@@ -0,0 +1,98 @@
+    qs = View.objects.all().values('id', 'project_id', 'data')
+
+    updated = 0
+    for row in qs.iterator(chunk_size=1000):
```
isn't there no point in using `iterator` now?

hakan458 replied:
oh thats right lmao

jombooth commented on label_studio/data_manager/migrations/0016_migrate_agreement_selected_annotators_to_unique.py:
```diff
@@ -0,0 +1,98 @@
+            continue
+
+        # Prepare new data JSON
+        new_data = dict(data)
+        new_agreement = dict(agreement)
+        new_agreement['annotators'] = list(new_annotators)
```
you can use `deepcopy` on `data` then update annotators in the result

jombooth commented on label_studio/data_manager/migrations/0016_migrate_agreement_selected_annotators_to_unique.py:
```diff
@@ -0,0 +1,98 @@
+        updated += 1
+
+    if updated:
+        print(f'{migration_name} Updated {updated} View rows')
```
is it worth logging the IDs as well? Not sure how many we have, but this might be useful if we need to debug after the fact

jombooth commented on label_studio/data_manager/migrations/0016_migrate_agreement_selected_annotators_to_unique.py:
```diff
@@ -0,0 +1,98 @@
+                .distinct()
+            )
+            # Normalize to unique ints
+            unique_ids = {int(u) for u in unique_ids}
```
It seems a bit redundant to both use `distinct` and a set to get unique ids

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17991622496

jombooth commented on label_studio/projects/api.py:
```diff
@@ -920,3 +922,38 @@ def delete(self, request, *args, **kwargs):
         return Response(data=count)
+
+
+@method_decorator(
+    name='get',
+    decorator=extend_schema(
+        tags=['Projects'],
+        summary='List annotators for project',
+        description='Return users who have submitted annotations in the specified project.',
+        responses={
+            200: OpenApiResponse(
+                description='List of annotator users',
+                response=UserSimpleSerializer(many=True),
+            )
+        },
+        extensions={
+            'x-fern-sdk-group-name': 'projects',
+            'x-fern-sdk-method-name': 'annotators',
+            'x-fern-audiences': ['public'],
+        },
+    ),
+)
```
please add a quick test for this once Spectacular / fern finish their work

hakan458 replied:
yep I will

jombooth commented on label_studio/projects/api.py:
```diff
@@ -920,3 +922,38 @@ def delete(self, request, *args, **kwargs):
+    name='get',
+    decorator=extend_schema(
+        tags=['Projects'],
+        summary='List annotators for project',
+        description='Return users who have submitted annotations in the specified project.',
```
```suggestion
        summary='List unique annotators for project',
        description='Return unique users who have submitted annotations in the specified project.',
```

jombooth commented on label_studio/projects/api.py:
```diff
@@ -920,3 +922,38 @@ def delete(self, request, *args, **kwargs):
+            )
+        },
+        extensions={
+            'x-fern-sdk-group-name': 'projects',
+            'x-fern-sdk-method-name': 'annotators',
```
```suggestion
            'x-fern-sdk-group-name': 'projects',
            'x-fern-sdk-method-name': 'list_unique_annotators',
```

jombooth reviewed (approved):
LG once feedback addressed

hakan458 committed:
update api description

Co-authored-by: Jo Booth <jo.m.booth@gmail.com>

hakan458 committed:
Update label_studio/projects/api.py

Co-authored-by: Jo Booth <jo.m.booth@gmail.com>

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17991901074

Commit:
updates

Commit:
Merge branch 'fb-utc-299' of github.com:heartexlabs/label-studio into fb-utc-299

Commit:
add sdk test

yyassi-heartex commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/18008420658)
> Successfully merged:  4 files changed, 42 insertions(+), 3 deletions(-)

yyassi-heartex committed:
Merge branch 'develop' into 'fb-utc-299'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18008420658

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18008642903

yyassi-heartex commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/18009284610)

yyassi-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18009284610

yyassi-heartex commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/18010152838)

yyassi-heartex committed:
Apply pre-commit linters

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18010152838

mcanu commented on label_studio/data_manager/migrations/0016_migrate_agreement_selected_annotators_to_unique.py:
```diff
@@ -0,0 +1,97 @@
+        project_id = row['project_id']
+        data = row.get('data') or {}
+
+        agreement = data.get('agreement_selected')
+        if not isinstance(agreement, dict):
+            continue
```
I believe that you can filter in the queryset itself through the JSONFIeld, this should limit a lot the loaded View objects into memory

mcanu commented on label_studio/data_manager/migrations/0016_migrate_agreement_selected_annotators_to_unique.py:
```diff
@@ -0,0 +1,97 @@
+
+        # Update only the JSON field via update(); do not load model instance or call save()
+        View.objects.filter(id=view_id).update(data=new_data)
+        print(f'Updated View {view_id} agreement selected annotators to {list(new_annotators)}')
```
Lets use logger.info, not sure how print behaves in prod

mcanu commented on label_studio/data_manager/migrations/0016_migrate_agreement_selected_annotators_to_unique.py:
```diff
@@ -0,0 +1,97 @@
+    migration.save(update_fields=['status'])
+
+def forwards(apps, schema_editor):
+    start_job_async_or_sync(forward_migration)
```
I expect this to be a long running job, can we put this in the low pri queue?

mcanu commented on label_studio/projects/api.py:
```diff
@@ -920,3 +922,38 @@ def delete(self, request, *args, **kwargs):
+        },
+    ),
+)
+class ProjectAnnotatorsAPI(generics.RetrieveAPIView):
```
I feel we should paginate this endpoint. We are always adding non paginated user endpoints and it comes back to bite us in the future.

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18010479122

Commit:
Merge branch 'develop' of github.com:heartexlabs/label-studio into fb-utc-299

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18014299834

Commit:
updates

Commit:
Merge branch 'fb-utc-299' of github.com:heartexlabs/label-studio into fb-utc-299

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18014621852

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/18014691703

robot-ci-heartex committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio-enterprise/actions/runs/18015602705
